CBinding/Gui/CTextEditorExtension.cs

```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
 						
 							// Default indentation method
 							Editor.InsertAtCaret (Editor.EolMarker + indent + GetIndent(Editor, Editor.OffsetToLineNumber (line.Offset), lineCursorIndex));
-							
 							return false;
 						
 					}
 				}
 			}
-			
 			return base.KeyPress (descriptor);
 		}
 
+		/// <summary>
+		/// Determine if code completion should be invoked by pressed char's (key's) value.
+		/// </summary>
+		/// <returns><c>true</c>, if complete should be invoked, <c>false</c> otherwise.</returns>
+		/// <param name="pressed">Pressed.</param>
+		bool shouldCompleteOn(char pressed) {
+			switch (pressed) {
+			case '.':
+				return true;
+			case '(':
+				return true;
+			case ' ':
+				return true;
+			case '>':
+				if (previous == '-')
+					return true;
+				return false;
+			case ':':
+				if (previous == ':')
+					return true;
+				return false;
+			default:
+				return false;
+
+			}
+		}
+
+		/// <summary>
+		/// Handles the code completion async. Invoked automatically on every keypress.
+		/// </summary>
+		/// <returns>The code completion async.</returns>
+		/// <param name="completionContext">Completion context.</param>
+		/// <param name="completionChar">Completion char.</param>
+		/// <param name="token">Token.</param>
 		public override Task<ICompletionDataList> HandleCodeCompletionAsync (CodeCompletionContext completionContext, char completionChar, CancellationToken token = default(CancellationToken))
 		{
-			string lineText = Editor.GetLineText (completionContext.TriggerLine).TrimEnd();
-			
-			// If the line ends with a matched extension, invoke its handler
-			foreach (KeyValuePair<string, GetMembersForExtension> pair in completionExtensions) {
-				if (lineText.EndsWith(pair.Key)) {
-					lineText = lineText.Substring (0, lineText.Length - pair.Key.Length);
-					
-					int nameStart = lineText.LastIndexOfAny (allowedCharsMinusColon) + 1;
-					string itemName = lineText.Substring (nameStart).Trim ();
-					
-					if (string.IsNullOrEmpty (itemName))
-						return null;
-					
-					return Task.FromResult ((ICompletionDataList)pair.Value (this, pair.Key, itemName));
+			ICompletionDataList list = new CompletionDataList ();
+			if (shouldCompleteOn(completionChar)) {
+				CProject project = DocumentContext.Project as CProject;
```

Comment:
Don't use an `as` cast when you expect it to always succeed. `as`returns null if the cast fails, so could cause a `NullReferenceException` when you access the object's member. A 'hard' cast will give an `InvalidCastException` if it fails:

    var project = (CProject)DocumentContext.Project;